src/aleph/http/http2.clj

```diff
@@ -49,102 +91,356 @@
+                         (doto (DefaultHttp2ResetFrame. h2-error-code)
+                               (.stream stream))))
+
+
+;; TODO: make stackless? Or consider if/when we need an exception
+(defn ^:no-doc shutdown-frame->h2-exception
+  "Converts an Http2ResetFrame or Http2GoAwayFrame into an Http2Exception or
+   StreamException, respectively."
+  ^Http2Exception
+  [evt stream-id]
+  ;; Sadly no common error parent class for Http2ResetFrame and Http2GoAwayFrame
+  (let [error-code (if (instance? Http2ResetFrame evt)
+                     (.errorCode ^Http2ResetFrame evt)
+                     (.errorCode ^Http2GoAwayFrame evt))
+        h2-error (Http2Error/valueOf error-code)
+        msg (if (instance? Http2ResetFrame evt)
+              (str "Received RST_STREAM from peer, stream closing. Error code: " error-code ".")
+              (str "Received GOAWAY from peer, connection closing. Error code: " error-code ". Stream " stream-id " was not processed by peer."))]
+    (if (instance? Http2ResetFrame evt)
+      (stream-ex stream-id msg {:h2-error h2-error})
+      (conn-ex msg {:on-stream-id stream-id} h2-error))))
+
+(defn add-header
   "Add a single header and value. The value can be a string or a collection of
    strings.
 
-   Respects HTTP/2 rules. All headers will be made lower-case. Strips invalid
-   connection-related headers. Throws on nil header values. Throws if
+   Header names are converted to lower-case AsciiStrings. Values are left as-is,
+   but it's up to the user to ensure they are valid. For more info, see
+   https://www.rfc-editor.org/rfc/rfc9113.html#section-8.2.1
+
+   Respects HTTP/2 rules. All headers will be made lower-case. Throws on
+   invalid connection-related headers. Throws on nil header values. Throws if
    `transfer-encoding` is present, but is not equal to 'trailers'."
   [^Http2Headers h2-headers ^String header-name header-value]
-  (log/debug "Adding HTTP header" header-name ": " header-value)
 
+  ;; Also checked by Netty, but we want to avoid work, so we check too
   (if (nil? header-name)
-    (illegal-arg "Header name cannot be nil")
+    (throw-illegal-arg-ex "Header name cannot be nil")
 
     ;; Technically, Ring requires lower-cased headers, but there's no guarantee, and this is
     ;; probably faster, since most users won't be caching.
     (let [header-name (if-some [cached (.get cached-header-names header-name)]
                         cached
-                        (let [lower-cased (str/lower-case (name header-name))]
+                        (let [lower-cased (-> header-name (name) (AsciiString.) (.toLowerCase))]
                           (.put cached-header-names header-name lower-cased)
                           lower-cased))]
+      (log/trace (str "Adding HTTP header: " header-name ": " header-value))
+
       (cond
-        (nil? header-value)
-        (illegal-arg (str "Invalid nil value for header '" header-name "'"))
+        ;; Will be checked by Netty
```

Comment:
It can be removed then.

Reply:
Yeah, ok. But I'm leaving a note to the future, so nobody wonders why we're not checking it